core/shared/src/main/scala/scalaz/zio/internal/OneShot.scala

```diff
@@ -0,0 +1,43 @@
+// Copyright (C) 2018 John A. De Goes
+package scalaz.zio.internal
+
+/**
+ * A variable that can be set a single time. The synchronous,
+ * effectful equivalent of `Promise`.
+ */
+private[zio] class OneShot[A] private (@volatile var value: A) {
+
+  /**
+   * Sets the variable to the value. The behavior of this function
+   * is undefined if the variable has already been set.
+   */
+  final def set(v: A): Unit =
```

Comment:
since the contract of this type is dependent on `value` being `null`, should we prohibit `v` from being `null`, or maybe use `Option`?

Reply:
Good point. Given this is internal and used in performance-critical sections, I'd rather prohibit `v` from being `null` than degrade performance with an `Option`.